### ISSUE TITLE
[CI] Builds failing due to Gradle test executor crash

### ISSUE THREAD

**Issue by mark-vieira** [elastic/elasticsearch#52610]
We are seeing occasional instances of errors that look like this:

```
Execution failed for task ':client:rest-high-level:test'.
> Process 'Gradle Test Executor 260' finished with non-zero exit value 1
This problem might be caused by incorrect test process configuration.
Please refer to the test execution section in the User Manual at https://docs.gradle.org/6.2/userguide/java_testing.html#sec:test_execution
```

Looking at the console and daemon logs we see these exceptions:

```
org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/127.0.0.1:41042'.
	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:94)
	at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:268)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:830)
Caused by: java.lang.IllegalArgumentException
	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:72)
	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81)
	... 7 more
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [22a1b02e-8ab0-42a3-ae0d-07afbc5a0358 port:37833, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:127)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:68)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:73)
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:776)
	at java.base/sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:120)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
	... 5 more
```

The strange thing about these messages is that none of the ports mentioned here are either the daemon server nor the Gradle client. So who's talking to who here and where are they getting these ports from? My other though was that maybe one of these is an ES test cluster node but none of the testclusters are using these ports. Perhaps interner test clusters?

I'm going to reach out to the folks at Gradle to try and track this down.

**Comment by droberts195** [comment]
This is still happening intermittently.  The two cases today so far are:

* https://gradle-enterprise.elastic.co/s/xa7w6cwjryeoc
* https://gradle-enterprise.elastic.co/s/ig66b43ygoks6

**Comment by DaveCTurner** [comment]
Another one recently:

- https://gradle-enterprise.elastic.co/s/2ak5oprmccpa2

**Comment by jkakavas** [comment]
Another one on 7.x intake https://gradle-enterprise.elastic.co/s/xaekqb5ejkpmc

**Comment by tvernum** [comment]
Another (7.x)
https://gradle-enterprise.elastic.co/s/hkbqbkj7tl37g/console-log?task=:plugins:analysis-icu:test

```
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [80d32bcf-ecbf-415a-af8a-ba2e1f7bb7eb port:37669, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].	
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)	
	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)	
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:122)	
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:70)	
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:68)	
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:73)	
Caused by: java.net.ConnectException: Connection refused	
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)	
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)	
	at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:111)	
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)	
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
```

**Comment by romseygeek** [comment]
Another (master)
https://gradle-enterprise.elastic.co/s/ul2coyq3v3d22/failure

**Comment by jaymode** [comment]
Another occurrence on 7.x: https://gradle-enterprise.elastic.co/s/sjj3dzchr5nau

```
Unexpected exception thrown.
org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/127.0.0.1:33632'.
        at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:94)
        at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.IllegalArgumentException: 
        at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:72)
        at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
        at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81)
        at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.lang.Thread.run(Thread.java:832)
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [1c684671-635a-4952-b141-a834070f48cf port:38057, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:123)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
	at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:111)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
	... 5 more
```

**Comment by rjernst** [comment]
@mark-vieira Any suggestion on how we can track down these failures? They appear to still be happening. Perhaps @breskeby could take a look?

**Comment by breskeby** [comment]
Do we have any way to reproduce one of those mentioned cases? I can have a closer look to figure out what's the root cause but it's hard with no way of reproducing and it seems they occurre quite rarely (but enough to be distrusting. If you see those, please link a gradle enterprise link to the according Build.

**Comment by rjernst** [comment]
@breskeby I believe this only happens in CI. I have never seen it locally. Here is one of those failures from today:
https://gradle-enterprise.elastic.co/s/wr2dr3w2hqcos

**Comment by mark-vieira** [comment]
My current guess is that we have internal cluster tests spinning nodes up on ports conflicting with either the daemon or test executor. I tweaked the way we capture the build logs for the GCP upload to include the full junit xml reports (which include stdout). The intention was to see if I can find the daemon/worker port anywhere in those logs. I haven't had a chance to actually go back and investigate after adding the logging though.

**Comment by breskeby** [comment]
I doubt its related to conflicts of ports here. My best guess at the moment is that something in the tests crashes the test executor jvm (e.g. calling System.exit(1) explicitly). Another potential issue might be tests conflicting or using a custom SecurityManager. Looking at some test fixtures I see some explicit System.exit(1) calls in there. I suggest we rework the fixtures to not use System.exit(1). I will look into this

**Comment by mark-vieira** [comment]
When I initially looked at the daemon logs there were messages about failing to communicate to the test worker. We had an identical issue with a regression in a recent Gradle release regarding local interface binding.

**Comment by breskeby** [comment]
We might see similar failures with different root causes the. With the latest build Ryan mentioned, the daemon logs look fine

**Comment by tvernum** [comment]
Another (7.x)
https://gradle-enterprise.elastic.co/s/hni5hb2ev37lg/console-log?task=:modules:reindex:test

**Comment by droberts195** [comment]
Another case in https://gradle-enterprise.elastic.co/s/koolcoru744pi

```
Unexpected exception thrown.
org.gradle.internal.remote.internal.MessageIOException: Could not write '/127.0.0.1:50596'.
	at org.gradle.internal.remote.internal.inet.SocketConnection.flush(SocketConnection.java:140)
	at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionDispatch.run(MessageHub.java:337)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.io.IOException: Connection reset by peer
	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:62)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:113)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:58)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:50)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:484)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeWithNonBlockingRetry(SocketConnection.java:279)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeBufferToChannel(SocketConnection.java:267)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.flush(SocketConnection.java:261)
	at org.gradle.internal.remote.internal.inet.SocketConnection.flush(SocketConnection.java:138)
	... 7 more
```

**Comment by droberts195** [comment]
Another case in https://gradle-enterprise.elastic.co/s/uo3zx5kuf6yak

```
* What went wrong:
Execution failed for task ':x-pack:plugin:ilm:internalClusterTest'.
> Process 'Gradle Test Executor 803' finished with non-zero exit value 1
  This problem might be caused by incorrect test process configuration.
  Please refer to the test execution section in the User Manual at https://docs.gradle.org/6.6.1/userguide/java_testing.html#sec:test_execution
```

**Comment by markharwood** [comment]
Another case in https://gradle-enterprise.elastic.co/s/ez5axqidlgvxk
7.9 CI LINUX elastic+elasticsearch+7.9+multijob-unix-compatibility os=debian-9&&immutable

	org.gradle.internal.remote.internal.ConnectException: Could not connect to server [06833148-67c0-4c86-8dd4-326b5691c2ca port:42837, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].	
		at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)	
		at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)	
		at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:123)	
		at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)	
		at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)	
		at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)	
	Caused by: java.net.ConnectException: Connection refused	
		at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)	
		at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)	
		at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:111)	
		at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)	
		at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)	
		... 5 more

**Comment by przemekwitek** [comment]
Yet another case in https://gradle-enterprise.elastic.co/s/cayul3sa6ufow
FAILURE elastic/elasticsearch#7.10 55e79dd release-tests - 20201207090007-9D5447B0

```
10:03:02 org.gradle.internal.remote.internal.ConnectException: Could not connect to server [bf866115-6be4-4209-9610-b45749518015 port:37991, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
10:03:02 	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
10:03:02 	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
10:03:02 	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:123)
10:03:02 	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
10:03:02 	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
10:03:02 	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
10:03:02 Caused by: java.net.ConnectException: Connection refused
10:03:02 	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
10:03:02 	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
10:03:02 	at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:111)
10:03:02 	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
10:03:02 	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
10:03:02 	... 5 more
10:03:02 Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
10:03:12 Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
```

**Comment by nik9000** [comment]
Another one: https://gradle-enterprise.elastic.co/s/2uspiahitbuwm

```
Unexpected exception thrown.
org.gradle.internal.remote.internal.MessageIOException: Could not write '/127.0.0.1:51968'.
        at org.gradle.internal.remote.internal.inet.SocketConnection.flush(SocketConnection.java:140)
        at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionDispatch.run(MessageHub.java:337)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.lang.Thread.run(Thread.java:832)
Caused by: java.io.IOException: Connection reset by peer
        at sun.nio.ch.FileDispatcherImpl.write0(Native Method)
        at sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:62)
        at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:113)
        at sun.nio.ch.IOUtil.write(IOUtil.java:58)
        at sun.nio.ch.IOUtil.write(IOUtil.java:50)
        at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:484)
        at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeWithNonBlockingRetry(SocketConnection.java:279)
        at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeBufferToChannel(SocketConnection.java:267)
        at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.flush(SocketConnection.java:261)
        at org.gradle.internal.remote.internal.inet.SocketConnection.flush(SocketConnection.java:138)
        at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionDispatch.run(MessageHub.java:337)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.lang.Thread.run(Thread.java:832)
Unexpected exception thrown.
org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/127.0.0.1:51968'.
        at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:94)
        at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.IllegalArgumentException: 
        at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:72)
        at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
        at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81)
        at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.lang.Thread.run(Thread.java:832)
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [f54c9ebc-e5f7-4052-a755-33f1087fde35 port:34431, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:123)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:715)
	at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:111)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
	... 5 more
```

**Comment by probakowski** [comment]
Another one: https://gradle-enterprise.elastic.co/s/kcx3euig3ntfc

```
org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/127.0.0.1:35914'. |  
-- | --
  | at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:94) |  
  | at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270) |  
  | at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64) |  
  | at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48) |  
  | at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) |  
  | at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) |  
  | at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56) |  
  | at java.lang.Thread.run(Thread.java:832) |  
  | Caused by: java.lang.IllegalArgumentException: |  
  | at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:72) |  
  | at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52) |  
  | at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81) |  
  | at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270) |  
  | at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64) |  
  | at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48) |  
  | at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) |  
  | at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) |  
  | at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56) |  
  | at java.lang.Thread.run(Thread.java:832) |  
  | org.gradle.internal.remote.internal.ConnectException: Could not connect to server [cd35320b-fcb9-4dfd-a4fe-65d6fd5f2c47 port:36481, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1]. |  
  | at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67) |  
  | at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36) |  
  | at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:123) |  
  | at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71) |  
  | at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69) |  
  | at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74) |  
  | Caused by: java.net.ConnectException: Connection refused |  
  | at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) |  
  | at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:776) |  
  | at java.base/sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:120) |  
  | at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81) |  
  | at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54) |  
  | ... 5 more


```

**Comment by mark-vieira** [comment]
This is definitely one of the worse non-test related build issues we have. Still at a loss for ideas as to what is causing this or how best to mitigate.

**Comment by benwtrent** [comment]
3 weeks since the last comment about it happening. But here we are again: https://gradle-enterprise.elastic.co/s/5cmu3molujlvs

```
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [88033f5e-c366-491d-9a08-edb78d436aec port:46627, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:123)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:776)
	at java.base/sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:120)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
	... 5 more
```

**Comment by mark-vieira** [comment]
@breskeby this is starting to be really problematic. We have at least a few of these every day. Any thoughts?

**Comment by breskeby** [comment]
No idea tbh what's causing this. I can invest some time next week to look into this. its hard to diagnose as it happens so irregular

**Comment by mark-vieira** [comment]
Or some other mitigation? It would be great if we could retry in this scenario but I think the issue is the test worker is dying.

**Comment by mark-vieira** [comment]
I'm wondering if we could take some inspiration from the test retry plugin to reexecute when we encounter a connect exception?

https://github.com/gradle/test-retry-gradle-plugin

**Comment by mark-vieira** [comment]
This should be addressed by https://github.com/elastic/elasticsearch/pull/71881.

**Comment by cbuescher** [comment]
@mark-vieira found this issue on test triage today googling the exception line, so maybe you want to take another look even though this is closed?

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+multijob+platform-support-unix/os=debian-9&&immutable/38/console

Build: https://gradle-enterprise.elastic.co/s/3zxtbqrj2zivg
```
15:15:10 Unexpected exception thrown.
15:15:10 org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/127.0.0.1:45320'.
15:15:10 	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:94)
15:15:10 	at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270)
15:15:10 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
15:15:10 	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
15:15:10 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
15:15:10 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
15:15:10 	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
15:15:10 	at java.base/java.lang.Thread.run(Thread.java:832)
15:15:10 Caused by: java.lang.IllegalArgumentException
15:15:10 	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:72)
15:15:10 	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
15:15:10 	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81)
15:15:10 	... 7 more
```

**Comment by benwtrent** [comment]
Happened again: https://gradle-enterprise.elastic.co/s/oerbc2dcndaf6

```
Unexpected exception thrown. |  
-- | --
  | org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/127.0.0.1:60286'. |  
  | at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:94) |  
  | at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270) |  
  | at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64) |  
  | at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48) |  
  | at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) |  
  | at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) |  
  | at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56) |  
  | at java.lang.Thread.run(Thread.java:831) |  
  | Caused by: java.lang.IllegalArgumentException: |  
  | at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:72) |  
  | at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52) |  
  | at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81) |  
  | at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270) |  
  | at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64) |  
  | at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48) |  
  | at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) |  
  | at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) |  
  | at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56) |  
  | at java.lang.Thread.run(Thread.java:831) |  
  | org.gradle.internal.remote.internal.ConnectException: Could not connect to server [db2adc07-afb3-450f-ad78-23396fb168e0 port:35289, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1]. |  
  | at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67) |  
  | at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36) |  
  | at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:123) |  
  | at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71) |  
  | at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69) |  
  | at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74) |  
  | Caused by: java.net.ConnectException: Connection refused |  
  | at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) |  
  | at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:715) |  
  | at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:111) |  
  | at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81) |  
  | at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54) |  
  | ... 5 more
```

**Comment by breskeby** [comment]
We saw something different but similar failure today here: https://gradle-enterprise.elastic.co/s/eicalr3x7xio4/console-log?task=:x-pack:plugin:spatial:test

**Comment by cbuescher** [comment]
And another another one on 7.x: https://gradle-enterprise.elastic.co/s/hwjqosndrc5ms/console-log?task=:x-pack:plugin:data-streams:test

**Comment by ywangd** [comment]
These failures today on 7.15 and 7.16 are of the same nature:
* https://gradle-enterprise.elastic.co/s/4vqvn654lfqpa
* https://gradle-enterprise.elastic.co/s/6weil6a2dc2ba

**Comment by tvernum** [comment]
Another
- https://gradle-enterprise.elastic.co/s/wp5tkjsk6ufts/console-log#L2441

**Comment by ywangd** [comment]
Failed on master intake https://gradle-enterprise.elastic.co/s/fgvnxyqde5day/console-log#L5001

**Comment by ywangd** [comment]
Failed for 7.16 arm64 https://gradle-enterprise.elastic.co/s/k7lmeapnpv7ya/console-log#L2682

**Comment by DaveCTurner** [comment]
[Here is another failure of this nature](https://gradle-enterprise.elastic.co/s/lwkqauz34576g), this time with a SEGV in the JVM:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000ffffa0996a10, pid=71459, tid=120356
#
# JRE version: OpenJDK Runtime Environment (21.0+35) (build 21+35-2513)
# Java VM: OpenJDK 64-Bit Server VM (21+35-2513, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# C  [libc.so.6+0x7aa10]  strlen+0x10
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /dev/shm/elastic+elasticsearch+main+multijob+platform-support-arm/server/build/testrun/internalClusterTest/hs_err_pid71459.log
[117.600s][warning][os] Loading hsdis library failed
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
```

**Comment by mark-vieira** [comment]
Where did you find that error output? I couldn't find anything relevant in the GCP upload.

**Comment by DaveCTurner** [comment]
https://gradle-enterprise.elastic.co/s/lwkqauz34576g/console-log?page=6#L5354

**Comment by mark-vieira** [comment]
> https://gradle-enterprise.elastic.co/s/lwkqauz34576g/console-log?page=6#L5354

How did I not see that? 🤦 

We can start by making sure we capture `hs_err_pid71459.log` file so we can see what's going on. I'll get that done.

**Comment by mark-vieira** [comment]
The last batch of these were caused by an interaction between Lucene and a JVM bug. That's be resolved so I'm going to close this for now. We can open a new issue if this starts happening again for some other cause.
